### PR TITLE
Implement some of the PCT uptake functions

### PR DESCRIPTION
**Pull request by dabreegster**
The uptake model will return a number between 0 and 1. Rather than completely filter out a route or increase counts by 1 for the route, we'll always increase counts by this probability. This PR is just defining some of the uptake formulas from https://github.com/ITSLeeds/pct/blob/HEAD/R/uptake.R

### CODE REVIEW

File: src/plugins/uptake.rs [hunk 1]
```diff
@@ -1,9 +1,69 @@
...
     }
+}
+
+// Everything below from
+// https://github.com/ITSLeeds/pct/blob/e630464efeaef539b18647b10745b863c9cd9948/R/uptake.R
+
+// TODO What does gradient represent -- an average or total or something over the entire route?
+// gradient should be in [0, 100]
+// This returns [0.0, 1.0]
+fn pct_gov_target(distance_meters: f64, gradient_percent: f64) -> f64 {
```
Comment (dabreegster) [inline src/plugins/uptake.rs:23]: @Robinlovelace (and @stholder3 FYI), I copied two methods from uptake.R. Does the below look correct to you? Is this the most recent version of the code? What other functions from that file might we want? Or will we be calculating some of these anew for safe routes to schools?

Comment (Robinlovelace) [reply src/plugins/uptake.rs:23]: Woohoo, this is an exciting update. Will take a look now Dustin!

File: src/plugins/uptake.rs [hunk 2]
```diff
@@ -1,9 +1,69 @@
...
+        + (d2 * distance_km.sqrt())
+        + (d3 * distance_km.powi(2))
+        + (h1 * gradient_percent)
+        + (i1 * distance_km * gradient_percent)
+        + (i2 * distance_km.sqrt() * gradient_percent);
+    inverse_logit(p)
+}
+
+fn pct_go_dutch(distance_meters: f64, gradient_percent: f64) -> f64 {
+    let alpha = -3.959 + 2.523;
```
Comment (Robinlovelace) [inline src/plugins/uptake.rs:46]:
```suggestion
    let alpha = -4.018 + 2.550;
```

etc from the 2020 variant

**Comment by Robinlovelace** [comment]
Looks good to me. I would do some checks to see that the results match this:

image

File: src/plugins/uptake.rs [hunk 3]
```diff
@@ -1,9 +1,69 @@
...
+    let i1 = 0.009394;
+    let i2 = -0.05135;
+
+    // TODO Why clamp to 30km?
+    let distance_km = (distance_meters / 1000.0).min(30.0);
+
+    let p = alpha
+        + (d1 * distance_km)
+        + (d2 * distance_km.sqrt())
+        + (d3 * distance_km.powi(2))
```
Comment (Robinlovelace) [inline src/plugins/uptake.rs:38]: Note: an unfortunate quirk of this distance^2 term is that for large values above ~30km pcycle can begin to increase! I mentioned this to Sam. Not sure how best to handle this, we use some conditional rule for large distances, ideally we need a totally new uptake function but this is the best representation of Dutch cycling levels we have so far.

**Review by Robinlovelace — APPROVED** [review]

File: src/plugins/uptake.rs [hunk 4]
```diff
@@ -65,5 +65,36 @@ fn pct_go_dutch(distance_meters: f64, gradient_percent: f64) -> f64 {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use fs_err::File;
+    use std::io::Write;
+
+    // Load the resulting CSV files with https://www.csvplot.com to manually check
+    #[test]
+    fn test_pct() {
```
Comment (Robinlovelace) [inline src/plugins/uptake.rs:83]: The tests that we never had in PCT : )

**Review by Robinlovelace — APPROVED** [review]

File: src/plugins/uptake.rs (hunk 4 above)
Comment (dabreegster) [reply src/plugins/uptake.rs:83]: Well, not exactly a test, just a raw to dump a CSV and plot it. Using the old versions (I haven't switched to 2020 yet):
[go_dutch.csv](https://github.com/dabreegster/routing-engines/files/12323533/go_dutch.csv)
[gov_target.csv](https://github.com/dabreegster/routing-engines/files/12323534/gov_target.csv)
https://www.csvplot.com/ displays the results. I haven't compared to the other plots yet

**Comment by Robinlovelace** [comment]
Looks good to me. 
image

**Comment by dabreegster** [comment]
Thanks for taking a look! I'll copy in 2020 numbers later. For now, this switches to accumulating the fractional uptake values.

**Pull request merged**